### PR TITLE
Backlinks

### LINKED ISSUE #607: Reverse lookup (Backlinks)
Hi guys,

I looked at the documentation and found support for linked queries. Is reverse lookup currently possible with the java version of Realm? I'm also using Realm on Cocoa which just added support for reverse lookup and wondered if there is already support for this feature in Java.

Thanks a lot!

### PR DESCRIPTION
**Pull request by cmelchior**
Closes #607 
Replaces #2881
Design PR: #2904 

This PR adds support for Backlinks or LinkingObjects.

-----
TODO
- [ ] ....
- [ ] Missing unit tests (???)
   - [ ] Add UT for `distinct` on backlinks
- [ ] Support backlinks in queries
- [ ] Website docs
  - [ ] Multiple links from instance A to instance B cause B to have multiple backlinks.
  - [ ] We do not, currently, support backlinks in libraries.
- [x] Finalize naming: `@Backlink` vs. `@LinkingObjects`
- [x] Make sure that backlink fields are ignored when importing from JSON 
- [ ] Decide how and if we should support backlinks across modules
---- 

@bmeike Creating this PR from your branch to make discussion easier. Please add any missing TODO's

### CODE REVIEW

**Comment by bmeike** [comment]
Backlinked fields are never added to the `ClassMetaData.fields` list, and so are not in the list of fields over which the method `emitCreateOrUpdateUsingJsonObject` iterates, when creating the `createOrUpdateUsingJsonObject` proxy method.

Added Unit Test to confirm.

**Comment by bmeike** [comment]
We are sticking with @LinkingObjects because iOS.

File: realm/realm-library/src/main/cpp/io_realm_internal_UncheckedRow.cpp [hunk 1]
```diff
@@ -375,3 +375,19 @@ JNIEXPORT jlong JNICALL Java_io_realm_internal_UncheckedRow_nativeGetFinalizerPt
...
+  (JNIEnv *env, jobject, jlong nativeRowPtr, jlong srcTablePtr, jlong srcColumnIndex) {
+
+    TR_ENTER_PTR(nativeRowPtr)
+    if (!ROW_VALID(env, ROW(nativeRowPtr)))
+        return -1;
+    try {
+        Row* row = ROW(nativeRowPtr);
+        Table* srcTable = TBL(srcTablePtr);
+        TableView backlinkView = row->get_table()->get_backlink_view(row->get_index(), srcTable, srcColumnIndex);
+        TableView* tv = new TableView(backlinkView);
```
Comment (bdash) [inline realm/realm-library/src/main/cpp/io_realm_internal_UncheckedRow.cpp]: `backlinkView` should be moved here.

**Comment by Zhuinden** [comment]
Can someone explain what `@LinkingObjects` means?

I mean, `@Backlink` obviously represents a relationship that links back to whoever points to this current RealmObject, right?

But what does **`@LinkingObjects`** mean? Especially considering even the internals call it a `backlink_view`.

**Comment by cmelchior** [comment]
@Zhuinden LinkingObjects mean "the objects linking to this object" .... Possible the word is just hard to parse for non-native English speakers because I feel the same as you. That said we discussed this internally and agreed that we think it important that Java/Cocoa align on the name and since no alternative to LinkingObject is clearly better, we just have to bite the bullet on this one.

E.g. the argument against Backlink is that it isn't a well-defined concept since it is normally called references when talking about object references, not links or forward links.

**Comment by Zhuinden** [comment]
@cmelchior ohhhhhhhhhhh.... it's not `objects that link (as in, "do linking"`, it's `the objects *linking* to this one`

Okay, I had to look at the word completely differently, now it makes sense. Thanks!

File: realm/realm-annotations-processor/src/main/java/io/realm/processor/Backlink.java [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+ * }
+ * </code>.
+ * <p>
+ * When managed, an instance X of a class with such a declaration will contain in the sourceField
+ * references to any instances of TargetClass whose targetField contains a reference to X.
+ * When managed, the sourceField cannot be assigned.  It can be queried normally.
+ * <p>
+ * When unmanaged, the sourceField is just another field: it can be set normally.  Managing an unmanaged
+ * instance of SourceClass destroys any previous contents of the backlinked field and replaces it with
+ * backlinks.  Unmanagning a managed instance of SourceClass nulls the backlinks field.
```
Comment (bmeike) [inline realm/realm-annotations-processor/src/main/java/io/realm/processor/Backlink.java]: Right.  Important distinction.

Comment (Zhuinden) [inline realm/realm-annotations-processor/src/main/java/io/realm/processor/Backlink.java]: `unmanagning` typo

File: realm/realm-annotations-processor/src/main/java/io/realm/processor/Backlink.java [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+ * <p>
+ * Note that, because subclassing subclasses of RealmObject is forbidden, so are constructs like:
+ * <code>RealmResults&lt;? extends Foos&lt;</code>
+ */
+final class Backlink {
+    private final VariableElement backlink;
+
+    /**
+     * The class containing the field <code>sourceField</code> with the backlink annotation
```
Comment (bmeike) [inline realm/realm-annotations-processor/src/main/java/io/realm/processor/Backlink.java]: ???? which?

File: realm/realm-annotations-processor/src/main/java/io/realm/processor/Backlink.java [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+    /**
+     * The name of the backlinked field, in sourceClass.
+     * The <code>RealmResults</code> field annotated with @LinkingObjects.
+     */
+    private final String sourceField;
+
+    /**
+     * The class to which the backlink, from sourceField, points
+     */
+    private final String targetClass;
```
Comment (bmeike) [inline realm/realm-annotations-processor/src/main/java/io/realm/processor/Backlink.java]: Sigh.  I am now using the exact same names as core: the tail of the link is called the "target" and the point is called the "source"

File: realm/realm-annotations-processor/src/test/resources/io/realm/AllTypesRealmProxy.java [hunk 5]
```diff
@@ -464,6 +464,26 @@ private void injectObjectContext() {
         }
     }
 
+    public RealmResults<some.test.AllTypes> realmGet$parentObjects() {
+        proxyState.getRealm$realm().checkIfValid();
+        BaseRealm realm = proxyState.getRealm$realm();
+        Row row = proxyState.getRow$realm();
+        Table srcTable = realm.getSchema().getTable(some.test.AllTypes.class);
+        Long idx = realm.getSchema().getSchemaForClass(some.test.AllTypes.class).getFieldIndex("columnObject");
```
Comment (bmeike) [inline realm/realm-annotations-processor/src/test/resources/io/realm/AllTypesRealmProxy.java]: Done.  Will add doc.

File: realm/realm-library/src/androidTest/java/io/realm/LinkingObjectsManagedTests.java [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+        AllJavaTypes child = realm.createObject(AllJavaTypes.class, 1);
+        AllJavaTypes parent = realm.createObject(AllJavaTypes.class, 2);
+        parent.setFieldObject(child);
+
+        try {
+            // Trying to set @LinkingObjects in managed mode is illegal
+            parent.setObjectParents(realm.where(AllJavaTypes.class).findAll());
+            fail();
+        }
+        catch (UnsupportedOperationException ignored) {
```
Comment (bmeike) [inline realm/realm-library/src/androidTest/java/io/realm/LinkingObjectsManagedTests.java]: fixed

File: realm/realm-library/src/androidTest/java/io/realm/LinkingObjectsManagedTests.java [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+        parent2.getFieldList().add(child);
+        realm.commitTransaction();
+
+        assertEquals(2, child.getListParents().size());
+    }
+
+    // Adding multiple list links creates multiple backlinks,
+    // even if the links are to a single object
+    @Test
+    public void multipleReferencesFromParentList() {
```
Comment (bmeike) [inline realm/realm-library/src/androidTest/java/io/realm/LinkingObjectsManagedTests.java]: Completely agree.  On the documentation list.

File: realm/realm-library/src/androidTest/java/io/realm/LinkingObjectsUnmanagedTests.java [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+        AllJavaTypes parent = realm.createObject(AllJavaTypes.class, 2);
+        parent.setFieldObject(child);
+        realm.commitTransaction();
+        assertEquals(1, parent.getFieldObject().getObjectParents().size());
+        assertEquals(child, parent.getFieldObject().getObjectParents().first());
+
+        AllJavaTypes unmanagedParent = realm.copyFromRealm(parent);
+
+        assertNull(unmanagedParent.getFieldObject().getObjectParents());
+    }
```
Comment (bmeike) [inline realm/realm-library/src/androidTest/java/io/realm/LinkingObjectsUnmanagedTests.java]: UT added

File: realm/realm-library/src/main/cpp/io_realm_internal_UncheckedRow.cpp (hunk 1 above)
Comment (bmeike) [inline realm/realm-library/src/main/cpp/io_realm_internal_UncheckedRow.cpp]: Huh?

Comment (bmeike) [inline realm/realm-library/src/main/cpp/io_realm_internal_UncheckedRow.cpp]: Huh?

File: realm-annotations/src/main/java/io/realm/annotations/LinkingObjects.java [hunk 9]
```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright 2016 Realm Inc.
```
Comment (bmeike) [inline realm-annotations/src/main/java/io/realm/annotations/LinkingObjects.java]: DOne

File: realm-annotations/src/main/java/io/realm/annotations/LinkingObjects.java [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+ * </pre>
+ *
+ * Linking objects have the following properties:
+ * <ul>
+ *     <li>The link is maintained by Realm and only works for managed objects.</li>
+ *     <li>They can be queried just like normal references.</li>
+ *     <li>They can be followed just like normal references.</li>
+ *     <li>They will be ignored when doing a `copyToRealm()`</li>
+ *     <li>They will be ignored when doing a `copyFromRealm()`</li>
+ * </ul>
```
Comment (bmeike) [inline realm-annotations/src/main/java/io/realm/annotations/LinkingObjects.java]: Added comment

File: realm/realm-annotations-processor/src/main/java/io/realm/processor/Backlink.java [hunk 11]
```diff
@@ -0,0 +1,237 @@
...
+ * <code>
+ * class SourceClass {
+ * // ...
+ * <p>
+ * {@literal @}LinkingObjects("targetField")
+ * RealmResults&lt;TargetClass&gt; sourceField;
+ * }
+ * </code>.
+ * <p>
+ * When managed, an instance X of a class with such a declaration will contain in the sourceField
```
Comment (bmeike) [inline realm/realm-annotations-processor/src/main/java/io/realm/processor/Backlink.java]: I think it is better now.

File: realm/realm-annotations-processor/src/main/java/io/realm/processor/Backlink.java (hunk 2 above)
Comment (bmeike) [inline realm/realm-annotations-processor/src/main/java/io/realm/processor/Backlink.java]: Yes!  Thoughto.  Fixed.

File: realm/realm-annotations-processor/src/main/java/io/realm/processor/Backlink.java (hunk 3 above)
Comment (bmeike) [inline realm/realm-annotations-processor/src/main/java/io/realm/processor/Backlink.java]: Done.

File: realm/realm-annotations-processor/src/main/java/io/realm/processor/Backlink.java (hunk 4 above)
Comment (bmeike) [inline realm/realm-annotations-processor/src/main/java/io/realm/processor/Backlink.java]: Now using the terms used by core: "source" is the point of the backlink, "target" is the tail.  That is, "source" is the parent, target the child.

File: realm/realm-annotations-processor/src/main/java/io/realm/processor/ClassMetaData.java [hunk 12]
```diff
@@ -70,27 +73,154 @@ public ClassMetaData(ProcessingEnvironment env, TypeElement clazz) {
...
         );
 
         for (Element element : classType.getEnclosedElements()) {
             if (element instanceof ExecutableElement) {
                 Name name = element.getSimpleName();
                 if (name.contentEquals("toString")) {
                     this.containsToString = true;
-                } else if (name.contentEquals("equals")) {
+                }
+                else if (name.contentEquals("equals")) {
```
Comment (bmeike) [inline realm/realm-annotations-processor/src/main/java/io/realm/processor/ClassMetaData.java]: Fixed.  Can we please get this into the auto-formatter?

File: realm/realm-annotations-processor/src/main/java/io/realm/processor/ClassMetaData.java [hunk 13]
```diff
@@ -70,27 +73,154 @@ public ClassMetaData(ProcessingEnvironment env, TypeElement clazz) {
...
+
+    public List<VariableElement> getFields() {
+        return Collections.unmodifiableList(fields);
+    }
+
+    public Set<Backlink> getBacklinkFields() {
+        return backlinks;
+    }
+
+    public String getGetter(String fieldName) {
```
Comment (bmeike) [inline realm/realm-annotations-processor/src/main/java/io/realm/processor/ClassMetaData.java]: Totally inherited that. Did not touch it.  Fixed.

File: realm/realm-annotations-processor/src/main/java/io/realm/processor/ClassMetaData.java [hunk 14]
```diff
@@ -70,27 +73,154 @@ public ClassMetaData(ProcessingEnvironment env, TypeElement clazz) {
...
+        return "realmGet$" + fieldName;
+    }
+
+    public String getSetter(String fieldName) {
```
Comment (bmeike) [inline realm/realm-annotations-processor/src/main/java/io/realm/processor/ClassMetaData.java]: Same.

File: realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProcessor.java [hunk 15]
```diff
@@ -245,4 +259,32 @@ private boolean createMediator(String simpleModuleName, Set<ClassMetaData> modul
 
         return true;
     }
+
+    // Because library classes are processed separately, there is no guarantee
```
Comment (bmeike) [inline realm/realm-annotations-processor/src/main/java/io/realm/processor/RealmProcessor.java]: Added task to comment

File: realm/realm-annotations-processor/src/test/resources/io/realm/AllTypesRealmProxy.java (hunk 5 above)
Comment (bmeike) [inline realm/realm-annotations-processor/src/test/resources/io/realm/AllTypesRealmProxy.java]: Yep.  WIll comment.

File: realm/realm-library/src/androidTest/java/io/realm/LinkingObjectsDynamicTests.java [hunk 16]
```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright 2016 Realm Inc.
```
Comment (bmeike) [inline realm/realm-library/src/androidTest/java/io/realm/LinkingObjectsDynamicTests.java]: Done

File: realm/realm-library/src/androidTest/java/io/realm/LinkingObjectsManagedTests.java (hunk 6 above)
Comment (bmeike) [inline realm/realm-library/src/androidTest/java/io/realm/LinkingObjectsManagedTests.java]: Fixed.  Let's get it into the formatter.

File: realm/realm-library/src/androidTest/java/io/realm/LinkingObjectsManagedTests.java (hunk 7 above)
Comment (bmeike) [inline realm/realm-library/src/androidTest/java/io/realm/LinkingObjectsManagedTests.java]: Will add distinct UT.

File: realm/realm-library/src/androidTest/java/io/realm/LinkingObjectsUnmanagedTests.java (hunk 8 above)
Comment (bmeike) [inline realm/realm-library/src/androidTest/java/io/realm/LinkingObjectsUnmanagedTests.java]: Done.  UT added.

File: realm/realm-library/src/main/cpp/io_realm_internal_UncheckedRow.cpp (hunk 1 above)
Comment (bdash) [inline realm/realm-library/src/main/cpp/io_realm_internal_UncheckedRow.cpp]: `TableView* tv = new TableView(std::move(backlinkView));`

It allows the state within `backlinkView` to be moved into the newly-allocated object, rather than having to create a copy of it only to then deallocate the original.

**Comment by bmeike** [comment]
Is there any way to change the source ("from") branch for this?  I am, as requested, now working on the branch 'feature/backlinks'

**Comment by bmeike** [comment]
Closing into #4219
